pkg/querier/distributor_queryable.go

```diff
@@ -38,46 +39,56 @@ type Distributor interface {
 	LabelValuesCardinality(ctx context.Context, labelNames []model.LabelName, matchers []*labels.Matcher) (uint64, *client.LabelValuesCardinalityResponse, error)
 }
 
-func newDistributorQueryable(distributor Distributor, iteratorFn chunkIteratorFunc, queryIngestersWithin time.Duration, logger log.Logger) QueryableWithFilter {
+func newDistributorQueryable(distributor Distributor, iteratorFn chunkIteratorFunc, cfgProvider distributorQueryableConfigProvider, logger log.Logger) QueryableWithFilter {
 	return distributorQueryable{
-		logger:               logger,
-		distributor:          distributor,
-		iteratorFn:           iteratorFn,
-		queryIngestersWithin: queryIngestersWithin,
+		logger:      logger,
+		distributor: distributor,
+		iteratorFn:  iteratorFn,
+		cfgProvider: cfgProvider,
 	}
 }
 
+type distributorQueryableConfigProvider interface {
+	QueryIngestersWithin(userID string) time.Duration
+}
+
 type distributorQueryable struct {
-	logger               log.Logger
-	distributor          Distributor
-	iteratorFn           chunkIteratorFunc
-	queryIngestersWithin time.Duration
+	logger      log.Logger
+	distributor Distributor
+	iteratorFn  chunkIteratorFunc
+	cfgProvider distributorQueryableConfigProvider
 }
 
 func (d distributorQueryable) Querier(ctx context.Context, mint, maxt int64) (storage.Querier, error) {
+	userID, err := tenant.TenantID(ctx)
+	if err != nil {
+		return nil, err
+	}
 	return &distributorQuerier{
```

Comment:
It is possible that even though `UseQueryable` returned `true`, now it would have returned `false`. (per-tenant value has changed)

Since `distributorQuerier` should use single value of `queryIngestersWithin`, we need to get it in this method. It means we could redo the check from `UseQueryable` and return noop querier if it now fails.

If `UseQueryable` returned `false` before, `Querier` will not be called, so we're good in that case.

Reply:
Added the check in https://github.com/grafana/mimir/pull/4287/commits/a63a36cad58ee98abedbcf8d4ed2e1d8f6e69ad2